app/wasm/query_plugin.go

```diff
@@ -41,6 +44,64 @@ func CustomQuerier(osmoKeeper ViewKeeper) func(ctx sdk.Context, request json.Raw
 				return nil, sdkerrors.Wrap(err, "osmo pool state query response")
 			}
 			return bz, nil
+		} else if contractQuery.SpotPrice != nil {
+			poolId := contractQuery.SpotPrice.Swap.PoolId
+			denomIn := contractQuery.SpotPrice.Swap.DenomIn
+			denomOut := contractQuery.SpotPrice.Swap.DenomOut
+			withSwapFee := contractQuery.SpotPrice.WithSwapFee
+
+			spotPrice, err := osmoKeeper.GetSpotPrice(ctx, poolId, denomIn, denomOut, withSwapFee)
+			if err != nil {
+				return nil, sdkerrors.Wrap(err, "osmo spot price query")
+			}
+
+			res := wasm.SpotPriceResponse{Price: spotPrice.String()}
+			bz, err := json.Marshal(res)
+			if err != nil {
+				return nil, sdkerrors.Wrap(err, "osmo spot price query response")
+			}
+			return bz, nil
+		} else if contractQuery.EstimatePrice != nil {
+			sender := "" // FIXME
```

Comment:
Huh? It seems they need a sender account with sufficient balance to do this query. Oooff...

Reply:
Let's just get the code looking good enough and merge it without tests.
Need to ask the Osmosis team about this and maybe change our API to include a sender.

Weird stuff though.
And let's test to ensure that queries really cannot change state when called by a contract (as the gamm code is changing state in the call afaik)

Reply:
All right. Added a "contract" sender field in https://github.com/confio/osmosis-bindings/pull/14. Will cut a osmosis-bindings v0.3.0, and change this.